src/Orleans.Runtime/Scheduler/SchedulerExtensions.cs

```diff
@@ -13,6 +13,7 @@ internal static Task<T> QueueTask<T>(this OrleansTaskScheduler scheduler, Func<T
                 {
                     try
                     {
+                        RequestContext.Clear();
```

Comment:
Is the request context valid after awaiting `QueueTask`?
```C#
RequestContext.Set("val", 5);
await scheduler.QueueTask(() => directory.LookupAsync(xxx), directory.SchedulingContext);
// Is RequestContext valid here?
Assert.Equal(RequestContext.Get("val"), 5); 
```

Reply:
Yes.  If the action is actually asynchronous, the logic you wrote should work.  However if it is not actually asynchronous the callers context may be cleared.  For instance:

     RequestContext.Set("val", 5);
     await Blarg();
     Assert.Equal(RequestContext.Get("val"), 5);   // will pass, because Blarg is asynchronous.
     await Blarg2();
     Assert.Equal(RequestContext.Get("val"), 5);  // will fail, because Blarg2 is not asynchronous.

     async Task Blarg()
     {
         RequestContext.Clear();
         await Task.Delay(1);
     }

     Task Blarg2()
     {
         RequestContext.Clear();
         return Task.CompletedTask;
     }

I'll add some testing to confirm.  If anyone knows how to force this behavior and have it not depend on whether the function is actually asynchronous or not, I'd love to here about it.  Best I could find is adding an await Task.Delay(Timespan.Zero) to the call. but I wasn't comfortable with that.

Reply:
I think you can test for a "pseudo-async" by examining the returned task object
```
var task = Blarg();
if (task.IsCompleted)
{
    // Blarg was not async but executed synchronously
}
else
{
   // Blarg is really async
}
await task;

```

Reply:
Understood, but I'm unclear how that test would help us.

Reply:
perhaps you could use this to detect if you overzealously deleted the callers requestcontext (i.e. if it is cleared before a synchronous call), and restore it.

Reply:
Hmm, something about the QueueTask is ensuring that it's always being handled correctly.  Added tests to confirm.  Will update PR with them.

Reply:
Ah ok, I believe this works because of copy-on-write semantics of the backing store for AsyncLocal. I was unaware before. I wonder if there's a need to clear the context in any other cases or if there's a way to move this clearing into the scheduling layer and have it more centralized (eg, clear when switching to a different scheduling context - not sure).